src/items/traits.md

```diff
@@ -95,6 +95,99 @@ Object safe traits can be the base trait of a [trait object]. A trait is
 * It must not have any associated constants.
 * All supertraits must also be object safe.
 
+When there isn't a `Self: Sized` bound on a method, the type of a method
+receiver must be one of the following types:
+
+* `&Self`
+* `&mut Self`
+* [`Box<Self>`]
+* [`Rc<Self>`]
+* [`Arc<Self>`]
+* [`Pin<P>`] where `P` is one of the object safe receiver types
+
+```rust
+# use std::rc::Rc;
+# use std::sync::Arc;
+# use std::pin::Pin;
+// Examples of object safe methods.
+trait TraitMethods {
+    fn by_ref(self: &Self) {}
+    fn by_ref_mut(self: &mut Self) {}
+    fn by_box(self: Box<Self>) {}
+    fn by_rc(self: Rc<Self>) {}
+    fn by_arc(self: Arc<Self>) {}
+    fn by_pin(self: Pin<&Self>) {}
+    fn with_lifetime<'a>(self: &'a Self) {}
+    fn nested_pin(self: Pin<Arc<Self>>) {}
+}
+# struct S;
+# impl TraitMethods for S {}
+# let t: Box<dyn TraitMethods> = Box::new(S);
+```
+
+```rust,compile_fail
+// These are object-safe, but cannot be dispatched on a trait object.
+trait NonDispatchable {
+    // Non-methods cannot be dispatched.
+    fn foo() where Self: Sized {}
+    // Self type isn't known until runtime.
+    fn returns(&self) -> Self where Self: Sized;
+    // `other` may be a different concrete type of the receiver.
+    fn param(&self, other: Self) where Self: Sized {}
```

Comment:
I can't find anything in the reference that mentions that this must be Sized for all traits. Am I missing something?

Reply:
Not sure I follow here.

Reply:
Why is this an error? ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=7af9cb55e569eb8d0e7f1c2a4f111ffd))

```rust
trait UnsizedSelf {
    fn param(&self, other: Self){}
}
```

Is there some rule listed somewhere that implies `other` must be sized?  I can't find it.

Reply:
Same result with:
```rust
trait UnsizedSelf { fn param(other: Self) {} }
```
or:
```rust
fn param<T: ?Sized>(other: T) {}
```
This is the general restriction that parameters, or local variables more generally, must have a known size.

Remember that traits are not `: Sized` by default.

By saying `where Self: Sized` you make `Self` have a known size for that function (meaning that it also is excluded from the VTable).

Reply:
More specifically, I was looking for a statement that says `Self` is `?Sized` for traits.  I guess the [Dynamically Sized Types](https://doc.rust-lang.org/nightly/reference/dynamically-sized-types.html#dynamically-sized-types) chapter covers that with the sentence "Unlike type parameters `Self: ?Sized` by default in trait definitions."  OK, nevermind, I just had a hard time finding that (I was expecting it in the traits chapter).

Reply:
Ah, I see. Well, glad you found it. =P